### PR TITLE
mxfdec: PHDR metadata track extraction

### PR DESCRIPTION
**Pull request by tomcrugnale-ssimwave**
Extract Prototype HDR metdata from MXF containers.

This format was developed jointly by Dolby and CineCert and is documented by Dolby as a means to carry Dolby Vision metadata.  

Global metadata (eg. mastering/target display information) is embedded once within an MXF file and is presented to the libav client as stream metadata.

Shot (or "frame") metadata is interleaved with the picture essence data and is presented to the to the libav client as a unique stream containing data packets.

### CODE REVIEW

File: libavformat/mxfdec.c [hunk 1]
```diff
@@ -2421,6 +2447,120 @@ static int mxf_add_metadata_stream(MXFContext *mxf, MXFTrack *track)
...
+    // Obtain the actual MXF track containing presumed Dolby Vision metadata, based on mapped track ID
+    for (size_t k = 0; k < mxf->metadata_sets_count; k++) {
+        MXFMetadataSet *metadata = mxf->metadata_sets[k];
+        if (metadata->type == Track && (track_id == ((MXFTrack*)(metadata))->track_id)) {
+            dovi_metadata_track = (MXFTrack*)metadata;
+            break;
+        }
+    }
+
+    av_log(NULL, AV_LOG_TRACE, "found in use Dolby Vision metadata track id %" PRIu64 "\n", track_id);
```
Comment (colekas-imax) [inline libavformat/mxfdec.c]: Does MXFContext have an AVContext that we can use here instead of NULL?

File: libavformat/mxfdec.c [hunk 2]
```diff
@@ -2421,6 +2447,120 @@ static int mxf_add_metadata_stream(MXFContext *mxf, MXFTrack *track)
...
+        return 0;
+    }
+
+    if (!(track->sequence = mxf_resolve_strong_ref(mxf, &track->sequence_ref, Sequence))) {
+        av_log(mxf->fc, AV_LOG_ERROR, "could not resolve track sequence strong ref\n");
+        return AVERROR_INVALIDDATA;
+    }
+
+    for (size_t j = 0; j < track->sequence->structural_components_count; j++) {
+        component = mxf_resolve_sourceclip(mxf, &track->sequence->structural_components_refs[j]);
```
Comment (colekas-imax) [inline libavformat/mxfdec.c:2512]: Is the idea that there's only 1 component within the refs?

File: libavformat/mxfdec.c [hunk 3]
```diff
@@ -2421,6 +2447,120 @@ static int mxf_add_metadata_stream(MXFContext *mxf, MXFTrack *track)
...
+
+    // Create a data stream which can be consumed by a client to obtain per-frame metadata
+    st = avformat_new_stream(mxf->fc, NULL);
+    if (!st) {
+        av_log(mxf->fc, AV_LOG_ERROR, "could not allocate Dolby Vision metadata stream\n");
+        return AVERROR(ENOMEM);
+    }
+
+    st->codecpar->codec_type = AVMEDIA_TYPE_DATA;
+    st->codecpar->codec_id = AV_CODEC_ID_NONE;
```
Comment (colekas-imax) [inline libavformat/mxfdec.c]: does it make sense to use bin data instead of none?

File: libavformat/mxfdec.c [hunk 4]
```diff
@@ -3158,6 +3298,53 @@ static int mxf_read_preface_metadata(void *arg, AVIOContext *pb, int tag, int si
...
+
+    read_res = avio_read(pb, mxf->dovi_global_metadata->data, size);
+    mxf->dovi_global_metadata->data[size] = '\0';
+
+    if (read_res >= 0) {
+        av_log(NULL, AV_LOG_TRACE, "PHDR global data: read %d bytes\n", read_res);
+        av_log(NULL, AV_LOG_TRACE, "PHDR global data body: %s", mxf->dovi_global_metadata->data);
+    }
+    else {
+        av_log(NULL, AV_LOG_TRACE, "Failed to read PHDR global data: result %d\n", read_res);
```
Comment (colekas-imax) [inline libavformat/mxfdec.c]: Do we want to free the av_mallocz stuff here?

File: libavformat/mxfdec.c [hunk 5]
```diff
@@ -4220,6 +4423,9 @@ static const AVOption options[] = {
     { "eia608_extract", "extract eia 608 captions from s436m track",
       offsetof(MXFContext, eia608_extract), AV_OPT_TYPE_BOOL, {.i64 = 0}, 0, 1,
       AV_OPT_FLAG_DECODING_PARAM },
+    { "dovi_metadata_extract", "extract Dolby Vision metadata",
+      offsetof(MXFContext, dovi_metadata_extract), AV_OPT_TYPE_BOOL, {.i64 = 1}, 0, 1,
```
Comment (colekas-imax) [inline libavformat/mxfdec.c]: We want to default this on for base FFmpeg?

**Review by colekas-ssim — APPROVED** [review]

File: libavformat/mxfdec.c (hunk 1 above)
Comment (tcrugnale-imax) [reply libavformat/mxfdec.c]: Good catch.  There are spots where I can reference an AVClass object for logging, so I'll update that.

File: libavformat/mxfdec.c (hunk 2 above)
Comment (tcrugnale-imax) [reply libavformat/mxfdec.c:2512]: Correct.

File: libavformat/mxfdec.c (hunk 3 above)
Comment (tcrugnale-imax) [reply libavformat/mxfdec.c]: Yeah, it's better than none.  There's an "AV_CODEC_ID_BINTEXT" but I think that actually maps to a real demuxer class, so I'll avoid that one.

File: libavformat/mxfdec.c (hunk 5 above)
Comment (tcrugnale-imax) [reply libavformat/mxfdec.c]: Initially my thinking was yes, but eia608_extract is disabled by default, and it's good to be consistent.

**Review by colekas-ssim — APPROVED** [review]

**Review by michellewen1516 — APPROVED** [review]

**Pull request merged**